File: shared/api-client.js
/**
 * Universal API client — works in any JS frontend (React, Vue, Svelte, vanilla).
 * Reads endpoints from api-contract.json so frontend stays in sync with backend.
 *
 * Usage:
 *   import api from '../shared/api-client';
 *   const items = await api.get('/items', { search: 'foo' });
 *   const newItem = await api.post('/items', { name: 'bar' });
 */

const BASE_URL = (typeof process !== 'undefined' && process.env?.NEXT_PUBLIC_API_URL)
  || (typeof process !== 'undefined' && process.env?.VITE_API_URL)
  || (typeof process !== 'undefined' && process.env?.REACT_APP_API_URL)
  || 'http://localhost:8000/api';

class ApiClient {
  constructor(baseURL = BASE_URL) {
    this.baseURL = baseURL.replace(/\/$/, '');
  }

  async request(method, path, { body, query, headers = {} } = {}) {
    let url = `${this.baseURL}${path}`;

    if (query) {
      const params = new URLSearchParams(
        Object.fromEntries(Object.entries(query).filter(([, v]) => v != null))
      );
      if (params.toString()) url += `?${params}`;
    }

    const opts = {
      method,
      headers: { 'Content-Type': 'application/json', ...headers },
    };

    if (body && method !== 'GET') {
      opts.body = JSON.stringify(body);
    }

    const res = await fetch(url, opts);

    if (!res.ok) {
      const error = new Error(`API ${method} ${path} failed: ${res.status}`);
      error.status = res.status;
      try { error.data = await res.json(); } catch {}
      throw error;
    }

    if (res.status === 204) return null;
    return res.json();
  }

  get(path, query)        { return this.request('GET', path, { query }); }
  post(path, body)        { return this.request('POST', path, { body }); }
  put(path, body)         { return this.request('PUT', path, { body }); }
  patch(path, body)       { return this.request('PATCH', path, { body }); }
  delete(path)            { return this.request('DELETE', path); }

  // Auth helper — call api.setToken(jwt) after login
  setToken(token) {
    this._token = token;
    const origRequest = this.request.bind(this);
    this.request = (method, path, opts = {}) => {
      opts.headers = { ...opts.headers, Authorization: `Bearer ${token}` };
      return origRequest(method, path, opts);
    };
  }
}

const api = new ApiClient();
export default api;
export { ApiClient };

File: shared/api-client.py
"""
Universal API client for Python backends that need to call other services,
or for scripts/notebooks that hit your own API.

Usage:
    from shared.api_client import api
    items = api.get('/items', params={'search': 'foo'})
    new_item = api.post('/items', json={'name': 'bar'})
"""

import os
import requests

BASE_URL = os.getenv("API_URL", "http://localhost:8000/api")


class ApiClient:
    def __init__(self, base_url=BASE_URL):
        self.base_url = base_url.rstrip("/")
        self.session = requests.Session()
        self.session.headers.update({"Content-Type": "application/json"})

    def request(self, method, path, **kwargs):
        url = f"{self.base_url}{path}"
        resp = self.session.request(method, url, **kwargs)
        resp.raise_for_status()
        if resp.status_code == 204:
            return None
        return resp.json()

    def get(self, path, params=None):
        return self.request("GET", path, params=params)

    def post(self, path, json=None):
        return self.request("POST", path, json=json)

    def put(self, path, json=None):
        return self.request("PUT", path, json=json)

    def patch(self, path, json=None):
        return self.request("PATCH", path, json=json)

    def delete(self, path):
        return self.request("DELETE", path)

    def set_token(self, token):
        self.session.headers["Authorization"] = f"Bearer {token}"


api = ApiClient()
